### ISSUE TITLE
OpenSeaDragonGL Web Demo Running Locally on Windows

### ISSUE THREAD

**Issue by zbackscatter** [thejohnhoffer/viaWebGL#3]
I'm trying to run the OpenSeaDragonGL Sobel demo on a Windows 7 workstation. I have OpenSeaDragon working but as soon as I attempt to run the OpenSeaDragonGL demo locally, it only displays the toolbar - no image. I was having issues with ES6 javascript compatibility so I changed the "for of" loops to ES5 compatibility. 

All scripts are loading, the browser shows the toolbars with a black background. Inspect shows the following errors:
"Failed to load resource: the server responded with a status of 404 (Not Found)"
"openseadragon.min.js:8 AJAX request returned 404: ../images/babel/babel.dzi" 

I checked my paths and the babel.dzi file exists. I'm at at loss as to what to try next. Thanks for your help in advance,
Tim

**Comment by zbackscatter** [comment]
One thing I forgot to mention is that I'm attempting to load the demo by running /demo/dzi/index.html locally.

**Comment by thejohnhoffer** [comment]
Warning: I'm easing myself into Windows. I'm downloading Visual Studio tonight and will try to serve the project from Visual Studio tomorrow. I've never user Visual Studio, and I don't know how to start a server from it. Is there a built-in simple http server? Most importantly, @zbackscatter, how are you specifically trying to serve the data from Visual Studio? 

If you're just opening the files locally _without an http server running_, I certainly wouldn't expect it to work as there would be problems including files in the html. If you're in any way **hosting** the files at localhost, then we should find a way to make it work.

Meanwhile- I've technically got the project running on Windows:

- I'm using the "Windows Subsystem For Linux" to run GNU bash v4.3.11 from cmd.exe.
- Using bash, I clone https://github.com/thejohnhoffer/viaWebGL under /mnt/c/Users/...
- I run python -m SimpleHTTPServer from the root directory of the repository.

![dzi demo from hosted by python on bash on windows][bash]

[bash]: http://img.hoff.in/viaGL/bashWin.png

**Comment by thejohnhoffer** [comment]
Also @zbackscatter, if you made changes to openseadragongl to remove unneeded ECMAScript 2015 [in line with fixing Issue 2][i2-0], you should definitely commit your changes and push them to the new the dev_master branch. Then I'll review them and pull them into the `master` branch. 

```
git pull origin dev_master
git checkout dev_master
git add -A; git commit -m 'no ecma6'
git push origin dev_master
```

[i2-0]: https://github.com/thejohnhoffer/viaWebGL/issues/2#issuecomment-273879382

**Comment by zbackscatter** [comment]
Hi John,

Nice work so far. I will take a look at what you’ve done to get it working under windows.

Visual Studio launches a simple web service called IISExpress when launching/debugging a web project from within Visual Studio. The web server in Windows is Internet Information Services (IIS). Its real easy to setup. It should be installed already on your windows machine. If not you can add it under Add/Remove Programs.

You can configure Visual Studio to use IIS but there’s no need. It will spin up the IIS Express process/web service to easily allow you to step through your client and server side code. During development we typically run and debug through the IISExpress web service and IIS for production. When I try to run under IISExpress web service, I get the problems which might be related to paths. I can try IIS tomorrow.

As far as Visual Studio goes, open it up and create a new project. Select  “ASP.NET Web Forms Application” for the simplest way to get up and running quickly.
[cid:image001.jpg@01D275A4.E1B30D60]
Select OK and let it setup. Then, right click on the solution or project name and select build as shown:
[cid:image002.png@01D275A4.E1B30D60]
Then run it by selecting the Internet Explorer button at the top:
[cid:image003.png@01D275A4.E1B30D60]
It should run.
You can then wire in your code to load openseadragon in the Default.aspx (HTML) page. Let me know what you need for help.
Thanks,
Tim


[AS&E]<http://www.as-e.com/>







 Timothy Houle
 Sr. Manager, Software Development
 829 Middlesex Turnpike  |  Billerica, MA 01821
 Office: 978-262-8700  |  Mobile: 978-835-5304
 thoule@as-e.com<mailto:thoule@as-e.com>
 [Twitter] <https://twitter.com/ase_detects> [LinkedIn] <https://www.linkedin.com/company/american-science-&-engineering-inc> [YouTube] <https://www.youtube.com/channel/UCOa_xvE1VN7MD_WHLqyw0FQ>


[http://www.as-e.com/uploads/main-images/signature_divider.gif]
 We are now part of the OSI Systems family of security companies: Rapiscan Systems, S2 Global, and AS&E.

**Comment by zbackscatter** [comment]
John,


Will do. I simply used the babel transpiler on the web to transpile the for if loops to ES5. I'll put the modified openseadragongl.js file with ES5 into the dev_master tomorrow morning or tonight if you need it sooner.

Cheers,


Tim




________________________________

**Comment by thejohnhoffer** [comment]
Tim,

It sounds like you've given me enough information to be able to replicate your setup to see if my webGL plugin for openSeadragon works with Visual Studio / ASP.NET / ISS Express. With all of these Microsoft frameworks, I'm lead to guess that it's simply a _backslash_ or _forwardslash_ issue. Since the backslash is the 'escape character' for javascript strings, a single backslash is written `\\`. We may need to change the dzi source folder marker from `/` to `\\` [on line 9 of demo/dzi/viewer.js][dzi] from this:

```javascript
this.source = '../images/babel/babel.dzi';
```

to this:


```javascript
this.source = '..\\images\\babel\\babel.dzi';
```

But I'm not sure about this. Theoretically servers on Windows should be able to handle the more widespread javascript convention of forward slashes for pathnames. 

[dzi]: https://github.com/thejohnhoffer/viaWebGL/blob/master/demo/dzi/viewer.js#L9

**Comment by zbackscatter** [comment]
John,

Interesting – excited to try out the proposed solution. I will try out immediately and let you know how things go.

Tim


[AS&E]<http://www.as-e.com/>







 Timothy Houle
 Sr. Manager, Software Development
 829 Middlesex Turnpike  |  Billerica, MA 01821
 Office: 978-262-8700  |  Mobile: 978-835-5304
 thoule@as-e.com<mailto:thoule@as-e.com>
 [Twitter] <https://twitter.com/ase_detects> [LinkedIn] <https://www.linkedin.com/company/american-science-&-engineering-inc> [YouTube] <https://www.youtube.com/channel/UCOa_xvE1VN7MD_WHLqyw0FQ>


[http://www.as-e.com/uploads/main-images/signature_divider.gif]
 We are now part of the OSI Systems family of security companies: Rapiscan Systems, S2 Global, and AS&E.

**Comment by zbackscatter** [comment]
Hi John,

Same thing. I tried and absolute and relative path inside viewer.js. there has got to be something else causing the TileSource not to load in IISExpress. If I install Python, did you say you got it working using the Python SimpleWebserver on Windows? Is that all I need to mimic what you did to get it running?

[cid:image001.png@01D27648.C96D44B0]




[AS&E]<http://www.as-e.com/>







 Timothy Houle
 Sr. Manager, Software Development
 829 Middlesex Turnpike  |  Billerica, MA 01821
 Office: 978-262-8700  |  Mobile: 978-835-5304
 thoule@as-e.com<mailto:thoule@as-e.com>
 [Twitter] <https://twitter.com/ase_detects> [LinkedIn] <https://www.linkedin.com/company/american-science-&-engineering-inc> [YouTube] <https://www.youtube.com/channel/UCOa_xvE1VN7MD_WHLqyw0FQ>


[http://www.as-e.com/uploads/main-images/signature_divider.gif]
 We are now part of the OSI Systems family of security companies: Rapiscan Systems, S2 Global, and AS&E.

**Comment by comrade2sun** [comment]
Hi
This may solve your issue.
To run the OpenSeaDragonGL Web Demo locally on Windows, just copy all the files in one folder, par example:
```
images
shaders
style
tools
index.html
main.js
viewer.js
```

Change all relative paths in the **_index.html_** like:
`src='../tools/osd/openseadragon.min.js'`
... and so on
to
`src='tools/osd/openseadragon.min.js'`

Do the same with all 'relative paths' in **_viewer.js_** and the demo will run locally in any browser without running a server on localhost. 
If you keep your file structure as mentioned, you are able to run your projects locally and/or on a remote server.

**Comment by thejohnhoffer** [comment]
@comrade2sun Thank you so much for the the help on this one. This sounds like the solution. 
@zbackscatter Have you tried [removing the `../` pathnames as suggested by @comrade2sun](https://github.com/thejohnhoffer/viaWebGL/issues/3#issuecomment-275065795)?

**Comment by zbackscatter** [comment]
I just tried comrade2sun's solution and it worked on FireFox! Thanks comrade2sun! Any ideas why only FireFox? Also, if I host the same directory structure (as specified by comrade2sun) thorugh a Microsoft IIS web service, it does not work for any browser, even FireFox.

**Comment by zbackscatter** [comment]
A 404 on a valid URL - Error on Chrome
![image](https://cloud.githubusercontent.com/assets/25230989/22319580/13921ab8-e352-11e6-908b-52e62e921364.png)

Error on FireFox - through web service
![image](https://cloud.githubusercontent.com/assets/25230989/22319657/bf3956ba-e352-11e6-86c9-f1d49e43bb71.png)


Here is my web service directory:
![image](https://cloud.githubusercontent.com/assets/25230989/22319618/5c3871a4-e352-11e6-9dd4-03cd738bb2e3.png)

**Comment by comrade2sun** [comment]
@zbackscatter 

> Any ideas why only FireFox?

I only can guess, it's because of the different <a href="https://en.wikipedia.org/wiki/JavaScript_engine#Active_projects">JavaScript engines</a> and maybe some security measures(?), in case it comes down to a local file system. Ironically IE11 is the only browser who doesn't care at all, but useless when it comes to this GSL shader stuff.
> A 404 on a valid URL - Error on Chrome

Double check the 'relative paths' in `viewer.js`

```javascript
SOBEL.Viewer = function() {
    // Needed constants
    this.tile_mode = 'tile-drawing';
    this.iconPrefix = 'images/icons/';
    this.source = 'images/babel/babel.dzi'; //check this !!!
    this.vShader = 'shaders/vertex/square.glsl';
    this.fShader = 'shaders/fragment/sobel3.glsl';
}
```

... and check your 'images' folder, if it containes the `babel.dzi` file and the `babel_files` folder and all its subfolders.

If everything is correct, you should be able to run your `index.html` **_in Firefox locally_**, par example: 
`file:///C:/inetpub/wwwroot/index.html`
or in **any[sic!]** other Browser via 
`http://localhost/index.html`

Good Luck

**Comment by comrade2sun** [comment]
@zbackscatter 
One more thought, why you may run in trouble with "Microsoft IIS web service":
Just rename the file 

> babel.dzi 

in 

> babel.xml 

and do the same in your 'viewer.js'
```javascript
this.source = 'images/babel/babel.xml'; 
```

I hope this will solve your issue

**Comment by zbackscatter** [comment]
comrade2sun, 
Thanks for the response. I changed the tileSource type to "image" and specified a JPG which worked fine in FireFox locally only. No chrome, or IE. 

```
        tileSources:
        {
            type: 'image',
            url: 'XrayImages/29200088.ZBV.28.Backscatter.0.jpg',
            buildPyramid: false
        },
```
I checked and verified permissions and verified its not '.xml' and verified all relative paths. FireFox seems to be OK. Chrome, I get:
![image](https://cloud.githubusercontent.com/assets/25230989/22390677/783dae8c-e4ba-11e6-92e5-0dd53aec5197.png)

I'll keep trying to hunt this down. If you think of anything else, lmk. Thanks for all your help so far to the both of you.

**Comment by zbackscatter** [comment]
I finally figured out the problem!! In IIS and IISExpress web servers, you must add the .glsl shader MIME entry or the shaders will not load properly. I tested with FF, CHrome, IE and Opera and they all work now.
![image](https://cloud.githubusercontent.com/assets/25230989/22394038/eca53a42-e4e2-11e6-929e-9433fa597cd3.png)

![image](https://cloud.githubusercontent.com/assets/25230989/22394048/28d3bd7c-e4e3-11e6-8626-8d5641ad07bd.png)
Thanks for all your help guys.
